src/signal_processing_algorithms/energy_statistics.py

```diff
@@ -0,0 +1,309 @@
+"""Energy statistics."""
+from dataclasses import dataclass
+from typing import List, Tuple, Union
+
+import numpy as np
+
+from signal_processing_algorithms.distance import get_distance_matrix
+
+
+@dataclass
+class EnergyStatisticsTestResult:
+    """Class for representing Energy Statistics and permutation test result."""
+
+    e: float
+    t: float
+    h: float
+    e_p: float
+    t_p: float
+    h_p: float
+
+
+class EnergyStatistics:
+    """
+    Energy Statistics of distributions x and y.
+
+    Calculates the following energy statistics of distributions x and y:
+    1. E-statistic
+    2. Test statistic
+    3. E-coefficient of inhomogeneity
+    """
+
+    def __init__(
+        self,
+        x: Union[List[float], List[List[float]], np.ndarray],
+        y: Union[List[float], List[List[float]], np.ndarray],
+    ) -> None:
+        """
+        Initialize.
+
+        :param x: A distribution which is an m x n array of m observations for n variables.
+        :param y: A distribution which is an l x n array of l observations for n variables.
+        """
+        self._x = get_valid_input(x)
+        self._y = get_valid_input(y)
+
+        if self._x.shape[1] != self._y.shape[1]:
+            raise ValueError(
+                "Number of variables per observation must be the same for both distributions!"
+            )
+
+        distances_within_x = get_distance_matrix(self._x, self._x)
+        distances_within_y = get_distance_matrix(self._y, self._y)
+        distances_between_xy = get_distance_matrix(self._x, self._y)
+
+        # E-statistic
+        self.e = self.get_e_from_distance_matrices(
+            distances_within_x, distances_within_y, distances_between_xy
+        )
+
+        self._n = len(self._x)  # Number of samples from X
+        self._m = len(self._y)  # Number of samples from Y
+
+        # Test statistic
+        self._t_coefficient = (self._n * self._m) / (self._n + self._m)
+        self.t = self.get_t_from_e(self._t_coefficient, self.e)
```

Comment:
Probably clearer to just combine these in to one line and remove `get_t_from_e`.

Reply:
Done.